solver/src/solver/matcha_solver/api.rs

```diff
@@ -0,0 +1,230 @@
+        }
+
+        let mut url = base_url
+            .join("/swap/v1/quote?")
+            .expect("unexpectedly invalid URL segment");
+        url.query_pairs_mut()
+            .append_pair("sellToken", &addr2str(self.sell_token))
+            .append_pair("buyToken", &addr2str(self.buy_token))
+            .append_pair("slippagePercentage", &self.slippage_percentage.to_string());
+        // I am not setting any affiliate address, in order to save gas
+        // .append_pair("affiliateAddress", "0xgp_address")
+        if let Some(amount) = self.sell_amount {
+            url.query_pairs_mut()
+                .append_pair("sellAmount", &amount.to_string());
+        }
+        if let Some(amount) = self.buy_amount {
+            url.query_pairs_mut()
+                .append_pair("buyAmount", &amount.to_string());
+        }
+        if let Some(skip_validation) = self.skip_validation {
+            url.query_pairs_mut()
+                .append_pair("skipValidation", &skip_validation.to_string());
+        }
+        url
+    }
+}
+
+/// A Matcha API swap response.
+#[derive(Clone, Derivative, Deserialize, PartialEq)]
+#[derivative(Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct SwapResponse {
+    #[serde(with = "u256_decimal")]
+    pub sell_amount: U256,
+    #[serde(with = "u256_decimal")]
+    pub buy_amount: U256,
+    pub allowance_target: H160,
+    #[serde(deserialize_with = "deserialize_decimal_f64")]
+    pub price: f64,
+    pub to: H160,
+    #[derivative(Debug(format_with = "debug_bytes"))]
+    pub data: Bytes,
+    #[serde(deserialize_with = "deserialize_decimal_u256")]
+    pub value: U256,
+}
+
+/// Mockable implementation of the API for unit test
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+pub trait MatchaApi {
+    async fn get_swap(&self, query: SwapQuery) -> Result<SwapResponse>;
+}
+
+/// Matcha API Client implementation.
+#[derive(Debug)]
+pub struct DefaultMatchaApi {
+    client: Client,
+    base_url: Url,
+}
+
+impl DefaultMatchaApi {
+    /// Create a new 1Inch HTTP API client with the specified base URL.
+    pub fn new(base_url: impl IntoUrl) -> Result<Self> {
+        Ok(Self {
+            client: default_http_client()?,
+            base_url: base_url.into_url()?,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl MatchaApi for DefaultMatchaApi {
+    /// Retrieves a swap for the specified parameters from the 1Inch API.
+    async fn get_swap(&self, query: SwapQuery) -> Result<SwapResponse> {
+        Ok(self
+            .client
+            .get(query.into_url(&self.base_url))
+            .send()
+            .await?
+            .json()
+            .await?)
+    }
+}
+
+impl Default for DefaultMatchaApi {
+    fn default() -> Self {
+        Self::new("https://api.0x.org/").expect("unexpected error parsing URL")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    #[ignore]
+    async fn test_api_e2e() {
```

Comment:
Maybe test both sides e2e (buy and sell order?)

My paraswap implementation had an issue on the buy order side which I could have caught with such a test

Reply:
I have the two tests that are very similar, just a little bit bigger:

-> solve_sell_order_on_matcha 
-> solve_buy_order_on_matcha

I think it is sufficient.